### PR TITLE
Ddcnl 3241

### PR DESCRIPTION
**Pull request by jpatton638**
- [X] Unit tests passing
- [X] Coverage reached
- [x] Acceptance tests passing
- [x] Reviewed

### CODE REVIEW

File: app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala [hunk 1]
```diff
@@ -16,33 +16,49 @@
...
     }
   }
+
+  private def extraHeaders(implicit hc: HeaderCarrier): HeaderCarrier =
+    hc.withExtraHeaders(
+      "Environment"   -> config.environment,
+      "Authorization" -> s"Bearer ${config.authorization}",
+      "Content-Type"  -> TaiConstants.contentType,
+      "CorrelationId" -> correlationId(hc)
+    )
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:58]: "X-Session-ID" -> hc.sessionId,
"X-Request-ID" -> hc.requestId

Comment (shikkhasingh) [reply app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:58]: @jpatton638 ^^

Comment (jpatton638) [reply app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:58]: @pascal22p What would you like in the event that either of these are empty?

**Review by shikkhasingh — CHANGES_REQUESTED** [review]

File: conf/application.conf [hunk 2]
```diff
@@ -133,17 +133,17 @@ tai {
   }
 
 
-  # Microservice specific config
+ # Microservice specific config
 
 auditing {
-enabled = true
+enabled = false
```
[the remark this answers is not in the document]
Comment (jpatton638) [reply conf/application.conf:139]: It was to clear up the logs while I was testing it, I'll change it back!

**Review by shikkhasingh — APPROVED** [review]

File: app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala (hunk 1 above)
Comment (struong) [reply app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:61]: @pascal22p what are you after here for empties?

File: app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala [hunk 3]
```diff
@@ -16,33 +16,54 @@
...
+  private def correlationId(hc: HeaderCarrier): String = {
+    val CorrelationIdPattern = """.*([A-Za-z0-9]{8}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}).*""".r
+    hc.requestId match {
+      case Some(requestId) =>
+        requestId.value match {
+          case CorrelationIdPattern(prefix) => prefix + "-" + randomUUID.toString.substring(24)
+          case _                            => randomUUID.toString
+        }
+      case _ => randomUUID.toString
     }
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:50]: Sorry this is obsolete now. The request id is going to change and won't be UUID anymore. Just use randomUUID.toString without deriving it from request_id

File: app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala [hunk 4]
```diff
@@ -16,33 +16,54 @@
...
   }
+
+  private def extraHeaders(implicit hc: HeaderCarrier): HeaderCarrier =
+    hc.withExtraHeaders(
+      "Environment"   -> config.environment,
+      "Authorization" -> s"Bearer ${config.authorization}",
+      "X-Session-ID"  -> hc.sessionId.fold(uuid)(_.value),
+      "X-Request-ID"  -> hc.requestId.fold(uuid)(_.value),
+      "Content-Type"  -> TaiConstants.contentType,
+      "CorrelationId" -> correlationId(hc)
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:60]: See above, just use randomUUID.toString

File: app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala [hunk 5]
```diff
@@ -16,33 +16,54 @@
...
+      case _ => randomUUID.toString
     }
   }
+
+  private def extraHeaders(implicit hc: HeaderCarrier): HeaderCarrier =
+    hc.withExtraHeaders(
+      "Environment"   -> config.environment,
+      "Authorization" -> s"Bearer ${config.authorization}",
+      "X-Session-ID"  -> hc.sessionId.fold(uuid)(_.value),
+      "X-Request-ID"  -> hc.requestId.fold(uuid)(_.value),
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/connectors/BbsiConnector.scala:58]: If there is no session id or request id do not make a new one, leave it empty.

**Comment by jpatton638** [comment]
@pascal22p ready for re-review

**Review by struong — APPROVED** [review]

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/master/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/master/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/master/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/master/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Pull request merged**